File: internal/ocr/normalizer.go
package ocr

import (
	"regexp"
	"strings"
	"unicode"
)

// Uzbek Standard Plate: 2 digits (region) + 1 letter + 3 digits + 2 letters, e.g. 01A123BC
var regexIndividualPlate = regexp.MustCompile(`^(\d{2})([A-Z])(\d{3})([A-Z]{2})$`)

// Uzbek Legal Entity Plate: 2 digits + 3 digits + 3 letters, e.g. 01123ABC
var regexLegalPlate = regexp.MustCompile(`^(\d{2})(\d{3})([A-Z]{3})$`)

// General fallback plate
var regexGeneralPlate = regexp.MustCompile(`^(\d{2})[A-Z0-9]{5,6}$`)

type NormalizationResult struct {
	Original   string
	Normalized string
	IsValid    bool
	PlateType  string // "INDIVIDUAL", "LEGAL", "GENERAL"
	ScoreBonus float64
}

func NormalizePlate(raw string) NormalizationResult {
	clean := strings.ToUpper(strings.TrimSpace(raw))
	// Remove all whitespace, hyphens, dots
	var sb strings.Builder
	for _, r := range clean {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			sb.WriteRune(r)
		}
	}
	cleaned := sb.String()

	if len(cleaned) < 7 || len(cleaned) > 9 {
		return NormalizationResult{
			Original:   raw,
			Normalized: cleaned,
			IsValid:    false,
			PlateType:  "INVALID",
			ScoreBonus: 0.0,
		}
	}

	// 1. Try direct match
	if regexIndividualPlate.MatchString(cleaned) {
		return NormalizationResult{
			Original:   raw,
			Normalized: cleaned,
			IsValid:    true,
			PlateType:  "INDIVIDUAL",
			ScoreBonus: 0.15,
		}
	}
	if regexLegalPlate.MatchString(cleaned) {
		return NormalizationResult{
			Original:   raw,
			Normalized: cleaned,
			IsValid:    true,
			PlateType:  "LEGAL",
			ScoreBonus: 0.15,
		}
	}

	// 2. Position-aware character ambiguity correction (0 vs O, 1 vs I, 5 vs S, 8 vs B)
	runes := []rune(cleaned)
	if len(runes) == 8 {
		// Individual format: [0,1] digits, [2] letter, [3,4,5] digits, [6,7] letters
		corr := make([]rune, 8)
		copy(corr, runes)

		corr[0] = forceDigit(corr[0])
		corr[1] = forceDigit(corr[1])
		corr[2] = forceLetter(corr[2])
		corr[3] = forceDigit(corr[3])
		corr[4] = forceDigit(corr[4])
		corr[5] = forceDigit(corr[5])
		corr[6] = forceLetter(corr[6])
		corr[7] = forceLetter(corr[7])

		candidate := string(corr)
		if regexIndividualPlate.MatchString(candidate) {
			return NormalizationResult{
				Original:   raw,
				Normalized: candidate,
				IsValid:    true,
				PlateType:  "INDIVIDUAL",
				ScoreBonus: 0.10,
			}
		}
	}

	if regexGeneralPlate.MatchString(cleaned) {
		return NormalizationResult{
			Original:   raw,
			Normalized: cleaned,
			IsValid:    true,
			PlateType:  "GENERAL",
			ScoreBonus: 0.05,
		}
	}

	return NormalizationResult{
		Original:   raw,
		Normalized: cleaned,
		IsValid:    false,
		PlateType:  "UNKNOWN",
		ScoreBonus: 0.0,
	}
}

func forceDigit(r rune) rune {
	switch r {
	case 'O', 'Q', 'D':
		return '0'
	case 'I', 'L':
		return '1'
	case 'Z':
		return '2'
	case 'S':
		return '5'
	case 'B':
		return '8'
	}
	return r
}

func forceLetter(r rune) rune {
	switch r {
	case '0':
		return 'O'
	case '1':
		return 'I'
	case '5':
		return 'S'
	case '8':
		return 'B'
	}
	return r
}

File: app/vision/normalizer.py
import re
from dataclasses import dataclass
from typing import Optional


# Regex for Uzbekistan License Plates
# 1. Standard Individual: 2 digits (01-95) + 1 letter (A-Z) + 3 digits + 2 letters (A-Z) -> 01A123BC (8 chars)
REGEX_INDIVIDUAL = re.compile(r"^(\d{2})([A-Z])(\d{3})([A-Z]{2})$")

# 2. Legal Entity: 2 digits (01-95) + 3 digits + 3 letters (A-Z) -> 01123ABC (8 chars)
REGEX_LEGAL = re.compile(r"^(\d{2})(\d{3})([A-Z]{3})$")

# 3. General Fallback: 2 digits + 5-6 alphanumeric characters (7-8 chars)
REGEX_GENERAL = re.compile(r"^(\d{2})[A-Z0-9]{5,6}$")


@dataclass
class NormalizationResult:
    original: str
    normalized: str
    is_valid: bool
    plate_type: str  # "INDIVIDUAL", "LEGAL", "GENERAL", "INVALID"
    score_bonus: float = 0.0


# Confusions mapping
LETTER_TO_DIGIT = {
    "O": "0", "Q": "0", "D": "0",
    "I": "1", "L": "1", "J": "1",
    "Z": "2",
    "S": "5",
    "B": "8",
    "G": "6",
}

DIGIT_TO_LETTER = {
    "0": "O",
    "1": "I",
    "2": "Z",
    "5": "S",
    "8": "B",
    "6": "G",
}


def force_digit(char: str) -> str:
    """Convert commonly confused letters to digits if position requires a digit."""
    return LETTER_TO_DIGIT.get(char, char)


def force_letter(char: str) -> str:
    """Convert commonly confused digits to letters if position requires a letter."""
    return DIGIT_TO_LETTER.get(char, char)


def clean_plate_text(raw: str) -> str:
    """Remove spaces, hyphens, dots and special characters, and convert to uppercase."""
    if not raw:
        return ""
    upper = raw.upper().strip()
    return "".join(c for c in upper if c.isalnum())


def normalize_plate(raw_text: str) -> NormalizationResult:
    """
    Normalize and validate Uzbekistan license plate text with position-aware OCR correction.
    """
    cleaned = clean_plate_text(raw_text)

    if len(cleaned) < 7 or len(cleaned) > 9:
        return NormalizationResult(
            original=raw_text,
            normalized=cleaned,
            is_valid=False,
            plate_type="INVALID",
            score_bonus=0.0,
        )

    # 1. Check direct match for Standard Individual Plate (01A123BC)
    if REGEX_INDIVIDUAL.match(cleaned):
        return NormalizationResult(
            original=raw_text,
            normalized=cleaned,
            is_valid=True,
            plate_type="INDIVIDUAL",
            score_bonus=0.15,
        )

    # 2. Check direct match for Legal Entity Plate (01123ABC)
    if REGEX_LEGAL.match(cleaned):
        return NormalizationResult(
            original=raw_text,
            normalized=cleaned,
            is_valid=True,
            plate_type="LEGAL",
            score_bonus=0.15,
        )

    chars = list(cleaned)

    # 3. Positional ambiguity correction for 8-character plates
    if len(chars) == 8:
        # Candidate A: Try Individual format (01 A 123 BC)
        # indices 0,1 -> digit; index 2 -> letter; indices 3,4,5 -> digit; indices 6,7 -> letter
        indiv = [
            force_digit(chars[0]),
            force_digit(chars[1]),
            force_letter(chars[2]),
            force_digit(chars[3]),
            force_digit(chars[4]),
            force_digit(chars[5]),
            force_letter(chars[6]),
            force_letter(chars[7]),
        ]
        candidate_indiv = "".join(indiv)
        if REGEX_INDIVIDUAL.match(candidate_indiv):
            return NormalizationResult(
                original=raw_text,
                normalized=candidate_indiv,
                is_valid=True,
                plate_type="INDIVIDUAL",
                score_bonus=0.10,
            )

        # Candidate B: Try Legal format (01 123 ABC)
        # indices 0,1 -> digit; indices 2,3,4 -> digit; indices 5,6,7 -> letter
        legal = [
            force_digit(chars[0]),
            force_digit(chars[1]),
            force_digit(chars[2]),
            force_digit(chars[3]),
            force_digit(chars[4]),
            force_letter(chars[5]),
            force_letter(chars[6]),
            force_letter(chars[7]),
        ]
        candidate_legal = "".join(legal)
        if REGEX_LEGAL.match(candidate_legal):
            return NormalizationResult(
                original=raw_text,
                normalized=candidate_legal,
                is_valid=True,
                plate_type="LEGAL",
                score_bonus=0.10,
            )

    # 4. Fallback General format (e.g. 7-character regional or trailers)
    if REGEX_GENERAL.match(cleaned):
        return NormalizationResult(
            original=raw_text,
            normalized=cleaned,
            is_valid=True,
            plate_type="GENERAL",
            score_bonus=0.05,
        )

    # 5. Invalid / Non-plate text (e.g. "ABC", "123", "HELLO", "XXXXXX")
    return NormalizationResult(
        original=raw_text,
        normalized=cleaned,
        is_valid=False,
        plate_type="INVALID",
        score_bonus=0.0,
    )
